h/services/bulk_annotation.py

```diff
@@ -0,0 +1,100 @@
+from typing import List
+
+import sqlalchemy as sa
+from sqlalchemy.orm import Session
+from sqlalchemy.sql import Select
+
+from h.db.query_filters import date_match
+from h.models import Annotation, AnnotationModeration, Group, User
+
+
+class BulkAnnotationService:
+    """A service for retrieving annotations for many users at once."""
+
+    def __init__(self, db_session: Session):
+        """Initialise the service."""
+
+        self._db = db_session
+
+    def annotation_search(
+        self, authority: str, audience: dict, updated: dict, limit=100000
+    ) -> List[Annotation]:
+        """
+        Get a list of annotations viewable by a particular set of users
+
+        :param authority: The authority to search by
+        :param audience: A specification of how to find the users (e.g)
+            {"username": [...]}
+        :param updated: A dict of operator to values to narrow down the updated
+            date
+        :param limit: A limit of results to generate
+        :raises BadDateFilter: For poorly specified date conditions
+        """
+        return (
+            self._db.execute(
+                self._search_query(authority, audience=audience, updated=updated).limit(
+                    limit
+                )
+            )
+            .scalars()
+            .all()
+        )
+
+    @classmethod
+    def _search_query(cls, authority, audience, updated) -> Select:
+        """Generate a query which can then be executed to find annotations."""
+
+        query = sa.select(Annotation)
+
+        # Updated
+        query = query.where(date_match(Annotation.updated, updated))
+
+        # Shared
+        query = query.where(Annotation.shared.is_(True))
+
+        # Deleted
+        query = query.where(Annotation.deleted.is_(False))
+
+        # Audience
+        query = query.join(Group, Annotation.groupid == Group.pubid).where(
+            Group.id.in_(cls._audience_groups_subquery(authority, audience))
+        )
+
+        # NIPSA
+        author_model = sa.orm.aliased(User, name="author")
+        query = query.join(
+            author_model,
+            author_model.username
+            == sa.func.split_part(sa.func.split_part(Annotation.userid, "@", 1), ":", 2)
+            and author_model.authority == authority,
+        ).where(author_model.nipsa.is_(False))
+
+        # Moderated
+        query = query.outerjoin(AnnotationModeration).where(
+            AnnotationModeration.id.is_(None)
+        )
```

Comment:
:+1: The query building seems pretty straightforward

Reply:
I guess we only need to retrieve certain columns from the DB, not entire rows. Might be a possible optimisation there?

Reply:
Yep, I'm not limiting the fields in any way yet.